File: ports/pyagentic/pyagentic/inference.py
"""Traditional ML / DL inference SPI (Tier 4) — ``Classifier`` and ``Scorer`` interfaces
plus two working, deterministic-by-default implementations and a ``ClassifierGuardrail``
that screens turns using a classifier instead of a regex.

- ``LexiconClassifier``: keyword-weighted bag-of-words classifier — no model, fully offline.
- ``EmbeddingClassifier``: nearest-centroid classifier over an ``Embedder`` — works offline
  with the hashing embedder and becomes a *real* semantic classifier when given a real
  embedder (Ollama/OpenAI via ``LiteLLMEmbedder``). This is the "real model opt-in" path:
  swap the embedder, keep the SPI.

Mirrors the Flink ``inference/`` SPI (Classifier/Scorer/guardrails) at a portable subset.
"""

from __future__ import annotations

import math
import re
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Protocol, Sequence, Tuple, runtime_checkable

from .embeddings import Embedder, HashingEmbedder
from .retrieval import cosine

_WORD = re.compile(r"[a-z0-9']+")


def _tokens(text: str) -> List[str]:
    return _WORD.findall((text or "").lower())


@dataclass(frozen=True)
class Classification:
    """A label + its confidence in [0,1], with the full per-label score map."""

    label: str
    score: float
    scores: Dict[str, float] = field(default_factory=dict)


@runtime_checkable
class Classifier(Protocol):
    def classify(self, text: str) -> Classification: ...


@runtime_checkable
class Scorer(Protocol):
    def score(self, text: str) -> float: ...


class LexiconClassifier:
    """Keyword-weighted classifier: each label has a list of keywords; the score for a
    label is the fraction of the text's tokens that hit that label's lexicon, softened so
    the per-label scores sum to 1. Deterministic, offline, no model."""

    def __init__(self, lexicon: Dict[str, Sequence[str]], default_label: str = "other") -> None:
        if not lexicon:
            raise ValueError("lexicon must have at least one label")
        self._lex = {label: {w.lower() for w in words} for label, words in lexicon.items()}
        self._default = default_label

    def _raw_scores(self, text: str) -> Dict[str, float]:
        toks = _tokens(text)
        if not toks:
            return {label: 0.0 for label in self._lex}
        n = len(toks)
        counts: Dict[str, float] = {}
        for label, words in self._lex.items():
            hits = sum(1 for t in toks if t in words)
            counts[label] = hits / n
        return counts

    def classify(self, text: str) -> Classification:
        raw = self._raw_scores(text)
        total = sum(raw.values())
        if total <= 0.0:
            scores = {label: 0.0 for label in raw}
            return Classification(self._default, 0.0, scores)
        scores = {label: v / total for label, v in raw.items()}
        label = max(scores, key=scores.get)
        return Classification(label, scores[label], scores)


class EmbeddingClassifier:
    """Nearest-centroid classifier over an ``Embedder``. ``fit`` averages the embeddings
    of each label's example texts into a centroid; ``classify`` returns the label whose
    centroid is most cosine-similar, with the similarity mapped to [0,1] as the score and
    a softmax over similarities as the per-label distribution. Real ML: with a real
    embedder this is a genuine semantic classifier; with the hashing embedder it's a
    deterministic offline default."""

    def __init__(self, embedder: Optional[Embedder] = None, temperature: float = 10.0) -> None:
        self._embedder = embedder or HashingEmbedder(256)
        self._temperature = temperature
        self._centroids: Dict[str, List[float]] = {}

    def fit(self, examples: Dict[str, Sequence[str]]) -> "EmbeddingClassifier":
        if not examples:
            raise ValueError("need at least one labeled example set")
        for label, texts in examples.items():
            texts = list(texts)
            if not texts:
                continue
            vecs = self._embedder.embed_batch(texts)
            dim = len(vecs[0])
            centroid = [0.0] * dim
            for v in vecs:
                for i, x in enumerate(v):
                    centroid[i] += x
            norm = float(len(vecs))
            self._centroids[label] = [x / norm for x in centroid]
        if not self._centroids:
            raise ValueError("no non-empty example sets")
        return self

    def classify(self, text: str) -> Classification:
        if not self._centroids:
            raise RuntimeError("classifier not fitted; call fit(...) first")
        vec = self._embedder.embed(text)
        sims: Dict[str, float] = {label: cosine(vec, c) for label, c in self._centroids.items()}
        # softmax over similarities for a proper distribution
        exps = {label: math.exp(self._temperature * s) for label, s in sims.items()}
        denom = sum(exps.values()) or 1.0
        scores = {label: e / denom for label, e in exps.items()}
        label = max(scores, key=scores.get)
        return Classification(label, scores[label], scores)


class ClassifierScorer:
    """Adapts a ``Classifier`` into a ``Scorer`` for one target label (its probability)."""

    def __init__(self, classifier: Classifier, label: str) -> None:
        self._classifier = classifier
        self._label = label

    def score(self, text: str) -> float:
        c = self._classifier.classify(text)
        return c.scores.get(self._label, c.score if c.label == self._label else 0.0)


class ClassifierGuardrail:
    """Guardrail that blocks when a classifier assigns a *blocked* label with confidence
    at or above ``threshold``. A learned/lexicon alternative to ``RegexGuardrail``."""

    def __init__(
        self,
        classifier: Classifier,
        blocked_labels: Sequence[str],
        threshold: float = 0.5,
        reason: str = "blocked by classifier policy",
        check_outputs: bool = False,
    ) -> None:
        self._classifier = classifier
        self._blocked = {label.lower() for label in blocked_labels}
        self._threshold = threshold
        self._reason = reason
        self._check_outputs = check_outputs

    def _hit(self, text: str) -> Optional[str]:
        if not text:
            return None
        c = self._classifier.classify(text)
        if c.label.lower() in self._blocked and c.score >= self._threshold:
            return f"{self._reason} ({c.label}={c.score:.2f})"
        return None

    def check_input(self, text: str) -> Optional[str]:
        return self._hit(text)

    def check_output(self, reply: str) -> Optional[str]:
        return self._hit(reply) if self._check_outputs else None

File: ports/experimental/go/core/inference.go
package core

import (
	"fmt"
	"math"
	"regexp"
	"strings"
)

// Traditional ML / DL inference SPI (Tier 4) — Classifier and Scorer interfaces plus two
// working, deterministic-by-default implementations and a ClassifierGuardrail that screens
// turns using a classifier instead of a regex.
//
//   - LexiconClassifier: keyword-weighted bag-of-words classifier — no model, fully offline.
//   - EmbeddingClassifier: nearest-centroid classifier over an Embedder — works offline with
//     the hashing embedder and becomes a real semantic classifier with a real embedder
//     (Ollama/OpenAI). This is the "real model opt-in" path: swap the embedder, keep the SPI.
//
// Mirrors the Flink inference/ SPI (Classifier/Scorer/guardrails) at a portable subset.

var wordRe = regexp.MustCompile(`[a-z0-9']+`)

func tokens(text string) []string {
	return wordRe.FindAllString(strings.ToLower(text), -1)
}

// Classification is a label + its confidence in [0,1], with the full per-label score map.
type Classification struct {
	Label  string
	Score  float64
	Scores map[string]float64
}

// Classifier assigns a label (with confidence) to a piece of text.
type Classifier interface {
	Classify(text string) Classification
}

// Scorer maps text to a single score in [0,1].
type Scorer interface {
	Score(text string) float64
}

// LexiconClassifier scores each label by the fraction of the text's tokens that hit that
// label's keyword set, normalised so the per-label scores sum to 1. Deterministic, offline.
type LexiconClassifier struct {
	lexicon      map[string]map[string]bool
	defaultLabel string
}

// NewLexiconClassifier builds a classifier from label -> keywords.
func NewLexiconClassifier(lexicon map[string][]string, defaultLabel string) *LexiconClassifier {
	lex := make(map[string]map[string]bool, len(lexicon))
	for label, words := range lexicon {
		set := make(map[string]bool, len(words))
		for _, w := range words {
			set[strings.ToLower(w)] = true
		}
		lex[label] = set
	}
	if defaultLabel == "" {
		defaultLabel = "other"
	}
	return &LexiconClassifier{lexicon: lex, defaultLabel: defaultLabel}
}

// Classify returns the best-matching label and a normalised score distribution.
func (c *LexiconClassifier) Classify(text string) Classification {
	toks := tokens(text)
	scores := make(map[string]float64, len(c.lexicon))
	for label := range c.lexicon {
		scores[label] = 0.0
	}
	if len(toks) == 0 {
		return Classification{Label: c.defaultLabel, Score: 0.0, Scores: scores}
	}
	n := float64(len(toks))
	total := 0.0
	for label, words := range c.lexicon {
		hits := 0
		for _, t := range toks {
			if words[t] {
				hits++
			}
		}
		scores[label] = float64(hits) / n
		total += scores[label]
	}
	if total <= 0.0 {
		return Classification{Label: c.defaultLabel, Score: 0.0, Scores: scores}
	}
	best, bestScore := c.defaultLabel, -1.0
	for label := range scores {
		scores[label] /= total
		if scores[label] > bestScore {
			best, bestScore = label, scores[label]
		}
	}
	return Classification{Label: best, Score: bestScore, Scores: scores}
}

// EmbeddingClassifier is a nearest-centroid classifier over an Embedder. Fit averages each
// label's example embeddings into a centroid; Classify returns the label whose centroid is
// most cosine-similar, mapped to a softmax distribution.
type EmbeddingClassifier struct {
	embedder    Embedder
	temperature float64
	centroids   map[string][]float64
}

// NewEmbeddingClassifier builds a classifier (nil embedder => hashing default; temp<=0 => 10).
func NewEmbeddingClassifier(embedder Embedder, temperature float64) *EmbeddingClassifier {
	if embedder == nil {
		embedder = NewHashingEmbedder(256)
	}
	if temperature <= 0 {
		temperature = 10.0
	}
	return &EmbeddingClassifier{embedder: embedder, temperature: temperature, centroids: map[string][]float64{}}
}

// Fit averages each label's example embeddings into a centroid.
func (c *EmbeddingClassifier) Fit(examples map[string][]string) (*EmbeddingClassifier, error) {
	for label, texts := range examples {
		if len(texts) == 0 {
			continue
		}
		vecs := c.embedder.EmbedBatch(texts)
		if len(vecs) == 0 || len(vecs[0]) == 0 {
			return nil, fmt.Errorf("embedder returned no vectors for label %q", label)
		}
		dim := len(vecs[0])
		centroid := make([]float64, dim)
		for _, v := range vecs {
			for i := 0; i < dim && i < len(v); i++ {
				centroid[i] += v[i]
			}
		}
		for i := range centroid {
			centroid[i] /= float64(len(vecs))
		}
		c.centroids[label] = centroid
	}
	if len(c.centroids) == 0 {
		return nil, fmt.Errorf("no non-empty example sets")
	}
	return c, nil
}

// Classify returns the nearest-centroid label with a softmax score distribution.
func (c *EmbeddingClassifier) Classify(text string) Classification {
	if len(c.centroids) == 0 {
		return Classification{Label: "other", Score: 0.0, Scores: map[string]float64{}}
	}
	vec := c.embedder.Embed(text)
	sims := make(map[string]float64, len(c.centroids))
	denom := 0.0
	for label, centroid := range c.centroids {
		sims[label] = math.Exp(c.temperature * Cosine(vec, centroid))
		denom += sims[label]
	}
	if denom == 0 {
		denom = 1.0
	}
	best, bestScore := "other", -1.0
	scores := make(map[string]float64, len(sims))
	for label, e := range sims {
		scores[label] = e / denom
		if scores[label] > bestScore {
			best, bestScore = label, scores[label]
		}
	}
	return Classification{Label: best, Score: bestScore, Scores: scores}
}

// ClassifierScorer adapts a Classifier into a Scorer for one target label.
type ClassifierScorer struct {
	classifier Classifier
	label      string
}

// NewClassifierScorer builds a scorer over a classifier for the given target label.
func NewClassifierScorer(classifier Classifier, label string) *ClassifierScorer {
	return &ClassifierScorer{classifier: classifier, label: label}
}

// Score returns the probability the classifier assigns to the target label.
func (s *ClassifierScorer) Score(text string) float64 {
	c := s.classifier.Classify(text)
	if v, ok := c.Scores[s.label]; ok {
		return v
	}
	if c.Label == s.label {
		return c.Score
	}
	return 0.0
}

// ClassifierGuardrail blocks when a classifier assigns a blocked label with confidence at
// or above threshold. A learned/lexicon alternative to RegexGuardrail.
type ClassifierGuardrail struct {
	classifier   Classifier
	blocked      map[string]bool
	threshold    float64
	reason       string
	checkOutputs bool
}

// NewClassifierGuardrail builds a classifier-backed guardrail.
func NewClassifierGuardrail(classifier Classifier, blockedLabels []string, threshold float64, reason string, checkOutputs bool) *ClassifierGuardrail {
	blocked := make(map[string]bool, len(blockedLabels))
	for _, l := range blockedLabels {
		blocked[strings.ToLower(l)] = true
	}
	if reason == "" {
		reason = "blocked by classifier policy"
	}
	return &ClassifierGuardrail{classifier: classifier, blocked: blocked, threshold: threshold, reason: reason, checkOutputs: checkOutputs}
}

func (g *ClassifierGuardrail) hit(text string) string {
	if text == "" {
		return ""
	}
	c := g.classifier.Classify(text)
	if g.blocked[strings.ToLower(c.Label)] && c.Score >= g.threshold {
		return fmt.Sprintf("%s (%s=%.2f)", g.reason, c.Label, c.Score)
	}
	return ""
}

// CheckInput screens the inbound text.
func (g *ClassifierGuardrail) CheckInput(text string) string { return g.hit(text) }

// CheckOutput screens the outbound reply when checkOutputs is set.
func (g *ClassifierGuardrail) CheckOutput(reply string) string {
	if g.checkOutputs {
		return g.hit(reply)
	}
	return ""
}
